### PR TITLE
PR #33 (saveMetaBoxes) | unit tests only

### PR DESCRIPTION
**Pull request by rgadon107**
This PR adds _unit tests_ for the `save_meta_boxes()` callback function in the Central Hub plugin. The _integration tests_ with WordPress for this callback will be submitted under a separate PR.

### CODE REVIEW

**Comment by rgadon107** [comment]
@hellofromtonya This morning I went back to the Brain Monkey documentation to review how to test WordPress' `has_action()` and `has_filter()` functions. The test method to test that `has_action( 'save_post', 'spiralWebDB\Metadata\save_meta_boxes' );` is true returns a `false` value. 

I changed the structure of the test method to set an expectation. With `has_action()`, one sets the expectation as in the following example: 

>Monkey\Actions\expectAdded( 'save_post' )
>  ->once()
>  ->with( 'spiralWebDB\Metadata\save_meta_boxes' )
>  ->andReturn(); 

The callback registered to the event to be mocked is passed as the argument. I'm still getting an error message on the CLI. It has to do with what the expectation is returning. 

Error message reads: 

>Brain\Monkey\Expectation\Exception\NotAllowedMethod: Bad usage of "andReturn" method: returning expectation can only be used for functions or applied filters expectations.

I traced the source of the message and the business logic that threw the Exception. 

I just don't know what's an acceptable return value (if any) for the expectation. I know that by default, `has_action()` will return `true`. I tried the running the expectation with and without `true`. Didn't make a difference. 

This is as far as I've gotten in resolving the error.

**Comment by hellofromtonya** [comment]
@rgadon107 Take a look at the [BrainMonkey - Testing Added Hooks docs](https://brain-wp.github.io/BrainMonkey/docs/wordpress-hooks-added.html). Scroll down to the "Just a couple of things..."

>* expectations must be set before the code to be tested runs: they are called "expectations" for a reason;
>* argument validation done using with(), validates hook arguments, not function arguments, it means what is passed to add_action() or add_filter() excluding hook name itself.

What's happening?

The file loads and then the callback is registered to the action. Then the test runs and the expectations are set. That means the expectations are set _after_ the callback has already been registered.

This means you can't use BrainMonkey in this edge case to test that the callback has been registered. You can do it in the integration test, but not in the unit test.

**Comment by rgadon107** [comment]
@hellofromtonya 
>(1) The file loads and then the callback is registered to the action. 

>(2) Then the test runs _and the expectations are set_. That means _the expectations are set **after the callback has already been registered**_.

Wrong order of setting the expectations. I get it now. 

Should I run an **integration test** for `has_action( 'save_post', 'spiralWebDB\MetaBoxes\save_meta_box' )`?

**Comment by rgadon107** [comment]
@hellofromtonya My latest commit transferred the test `test_function_should_register_to_action_hook_when_event_fires` from the unit to integration test class. The following test assert passes:

>(a) $this->assertTrue( has_action( 'save_post' ) );

The following test assert *does not pass*:

>(b) $this->assertTrue( has_action( 'save_post', 'spiralWebDB\Metadata\save_meta_boxes' ) );

I defer to your judgement as to whether the hook `save_post` should be tested or not. 

When I inspect the source throwing the fatal error in (b): 

>Fatal error: Default value for parameters with a class type hint can only be NULL in /app/public/wp-content/tests/vendor/symfony/yaml/Yaml.php on line 52

The argument in the Yaml.php file (L52) wants a file path. I don't understand how to fix the error.

File: mu-plugins/central-hub/src/meta-data/meta-box.php [hunk 1]
```diff
@@ -124,11 +124,15 @@ function is_okay_to_save_meta_box( $meta_box_key ) {
...
 	// Security check.
+	$nonce_name = "{$meta_box_key}_nonce_name";
+	if ( ! isset( $_POST[ $nonce_name ] ) ) {
+		return false;
+	}
+
 	return wp_verify_nonce(
-		$_POST[ $meta_box_key . '_nonce_name' ],
+		$_POST[ $nonce_name ],
 		$meta_box_key . '_nonce_action'
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/src/meta-data/meta-box.php]: @rgadon107 Let's go ahead and be consistent here by changing the string's construction to:

```
"{$meta_box_key}_nonce_action"
```

File: mu-plugins/central-hub/tests/phpunit/unit/meta-data/saveMetaBoxes.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		Monkey\Functions\expect( 'KnowTheCode\ConfigStore\getConfigParameter' )->never();
+		Monkey\Functions\expect( 'spiralWebDB\Metadata\save_custom_fields' )->never();
+		save_meta_boxes( 178 );
+	}
+
+
+	/**
+	 * Test save_meta_boxes() should not save when config meta box key does not begin with 'meta_box.'.
+	 */
+	public function test_should_not_save_when_config_meta_box_key_does_not_begin_with_metabox() {
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/meta-data/saveMetaBoxes.php]: @rgadon107 This test can be deleted. Why? It's covered by the above test. `get_meta_box_keys()` only returns keys that start with `meta_box.`. Therefore, if there are no keys that start with `meta_box.`, that means `get_meta_box_keys()` returns an empty array.

File: mu-plugins/central-hub/tests/phpunit/unit/meta-data/saveMetaBoxes.php [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+			->andReturn( false );
+		Monkey\Functions\expect( 'KnowTheCode\ConfigStore\getConfigParameter' )->never();
+		Monkey\Functions\expect( 'spiralWebDB\Metadata\save_custom_fields' )->never();
+		save_meta_boxes( 317 );
+	}
+
+	/**
+	 *  Test save_meta_boxes() should not save when the nonce key is not equal to the expected value.
+	 */
+	public function test_should_not_save_when_the_nonce_key_is_not_equal_to_the_expected_value() {
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/meta-data/saveMetaBoxes.php]: @rgadon107 This test is covered in the test below. If the value isn't expected, then the nonce check fails. Therefore, you can delete this test.

File: mu-plugins/central-hub/tests/phpunit/unit/meta-data/saveMetaBoxes.php [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	public function test_should_save_when_one_or_more_valid_post_meta_values_are_added() {
+		Monkey\Functions\expect( 'spiralWebDB\Metadata\get_meta_box_keys' )
+			->once()
+			->withNoArgs()
+			->andReturn( [ 'meta_box.members' ] );
+		Monkey\Functions\expect( 'spiralWebDB\Metadata\get_meta_box_id' )
+			->once()
+			->with( 'meta_box.members' )
+			->andReturn( 'members' );
+		Monkey\Functions\when( 'spiralWebDB\Metadata\is_okay_to_save_meta_box' )
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/meta-data/saveMetaBoxes.php]: @rgadon107 This mock should be an `expect` and turn `true`.

File: mu-plugins/central-hub/tests/phpunit/unit/meta-data/saveMetaBoxes.php [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+			->justReturn( 'wp_verify_nonce' );
+		Monkey\Functions\when( 'KnowTheCode\ConfigStore\getConfigParameter' )
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/meta-data/saveMetaBoxes.php]: @rgadon107 This mock should be an expect, as we expect it to run once and return the configuration.

File: mu-plugins/central-hub/tests/phpunit/unit/meta-data/saveMetaBoxes.php [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+			->justReturn( [
+				'role'            => 'Soprano',
+				'residence_city'  => 'Chicago',
+				'residence_state' => 'IL',
+				'tour_number'     => '3',
+			] );
+		Monkey\Functions\when( 'spiralWebDB\Metadata\save_custom_fields' )
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/meta-data/saveMetaBoxes.php]: @rgadon107 We expect this mock to run once, with the 3 arguments, and return `null`.
